doc/reading/divisibility-of-split.md

```diff
@@ -4,24 +4,249 @@
+The indices of the inputs of all these splits, `i0`, `i1`, and `i2`, could potentially run out of boundary.
+
+In order to maintain program semantics, the safest predication strategy is to predicate on all these three IterDomains:
+
+```python
+for i3 in range(2):
+    for i4 in range(2):
+        for i5 in range(2):
+            for i6 in range(4):
+                i1 = i3 * 2 + i4
+                i2 = i5 * 4 + i6
+                i0 = (i3 * 2 + i4) * 6 + (i5 * 4 + i6)
+                if i1 < 3 and i2 < 6 and i0 < 15:
+                    print(i0)
+```
+
+and this indeed works correctly.
+
+Can this be simplified? What if we just predicate `i0` and nothing else?
+We will get `0 1 2 3 4 5 6 7 6 7 8 9 10 11 12 13 12 13 14`.
+We do print the correct set of values, but we are printing some values multiple times.
+If all we care is to print the correct set of values, and we don't mind whether there are duplicates, this can be one strategy.
+But this is clearly not equivalent to the program prior to transformation.
+
+From the above example, we can see that overflow of an intermediate IterDomain does not necessarily results in an overflow of their ancestors.
+To maintain program semantics, the predicate we use must be logically equivalent to predicating all holes.
+
+The following theorems are useful tools to find mathematically simple predicates that are logically equivalent to predicating all holes:
+
+**Theorem 1** Suppose that there is a split `I1, I2 = Split(I0, N)`.
+Then "the index of `I0` is in bound" implies "the index of `I1` is in bound".
+
+<details>
+
+**<summary>Proof:</summary>**
+
+Suppose the index of `I1` is $i_1$, the index of `I2` is $i_2$, the extent of `I0` is $S$.
+The index of `I0` is then $i_0 = i_1 \times N + i_2$.
+"the index of `I0` is in bound" means $i_0 < S$.
+Because $i_2 \ge 0$,
+$$i_0 < S \implies i_1 \times N < S \implies i_1 < S \div N \implies i_1 < \mathrm{ceilDiv}(S, N)$$
+$\square$
+
+</details>
+
+The above theorem tells us that, if we are already predicating `I0`, then there is no need to additionally predicate `I1`.
+We can derive similar theorems for other IterDomain expressions:
+
+**Theorem 2** Suppose that there is a merge `I2 = Merge(I0, I1)`.
+Then "the index of `I0` is in bound" is equivalent to "the index of `I2` is in bound".
+
+<details>
+
+**<summary>Proof:</summary>**
+
+Suppose the index of `I2` is $i_2$, the extent of `I1` is $N$.
+Then the index of `I0` is $i_0 = i_2 / N$.
+Suppose that the extents of `I0` and `I2` are $P$ and $Q$, then $Q = N \times P$.
+"the index of `I0` is in bound" means $i_0 < P$, which is:
+$$i_2 / N < P$$
+According to "Rule 1" in `[Simplification of boolean predicates]` in `csrc/expr_simplifier.h`,
+(TODO: move this theorem to a md file)
+$$i_2 / N < P \Leftrightarrow i_2 < Q$$
+$\square$
+
+</details>
+
+**Theorem 3** Suppose that there is a resize `I1 = Resize(I0, L, R)`.
+Then "the index of `I0` is in bound" implies "the index of `I1` is in bound" if $R >= 0$.
```

Comment:
When you say `I0` is in bound, does that also mean `I0 >= -L`?

Reply:
Good catch! So far I have assumed that we only need to predicate upper bound, and all integers are automatically >= 0. Looks like this may not be the case for `Resize`. Let me thinking through it.

Reply:
I resolved this in https://github.com/NVIDIA/Fuser/pull/2060, it does make it more complicated, but the conclusion regarding TMA does not change.